Starting up a code chapter (#21)
fhdsl/Tools_for_Reproducible_Workflows_in_R: 06-code.Rmd

# Reproducible Code

```{r, include = FALSE}
ottrpal::set_knitr_image_path()
```

> ```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Learning Objectives are to be able to: Find and utilize packages and their functions. Recognize aspects of reproducible code: Readable, Efficient , and Consistent. Work on code in a way that iteratively encourages the reproducibility of your project"}

Reviewer: ```suggestion
```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Learning Objectives are to be able to: Find and utilize packages and their functions. Recognize aspects of reproducible code: Readable, Efficient, and Consistent. Work on code in a way that iteratively encourages the reproducibility of your project"}
```

_This is the beauty of reproducibility, if you or someone else makes great code, it can not only be re-run but it can be repurposed!_

### Tips for choosing packages to use:

- Does the package have easy to use documentation to help guide you on how to use it properly? If the package has underdeveloped documentation it may be difficult for you or other to understand and use properly.
- Is the code actively being developed or maintained? Packages that are no longer being maintained will likely deprecate more quickly rendering them unusable in the future.
- Is it a package that is commonly recognized by the community? Well recognized packages will be easier for others to comment on and help you with.

## The importance of iterative work

We've mentioned previously that reproducibility is iterative work. This way of working refers to code work as much as anything else. You won't ever right perfect code on the first try, instead aim for each chunk to work one step at a time. Once it is working, take a break (perhaps until the next day), then return to it and look for ways to polish it and make it more efficient.

> For [more tips on how to work read this blog](https://www.ccdatalab.org/blog/not-so-secret-sauce-for-efficient-workflows).

Reviewer: ```suggestion
[For more tips on how to work read this blog](https://www.ccdatalab.org/blog/not-so-secret-sauce-for-efficient-workflows).
```

Reviewer: maybe make the entire sentence a link? looked a tad funny with just the first word not part of the link

ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g221fb1e0594_0_6")
```

## Readable

Reproducible code is readable code. In order for other people to use your code, they will need to be able to understand it. Because of that, code being readable is more important than code being innovative or clever.

```{r, fig.align='center', out.width="100%", echo = FALSE, fig.alt= "Reproducible parrot is looking at their computer and says ‘It’s so easy for me to understand what is going on with this code because it is so readable and follows a style!’"}
ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g20ed8021e09_2_120")
```

### Well-documented

Reproducible code is well documented code! This includes (but isn't limited to):

- A README that can get individuals up to speed on the project quickly

> - Code and notebooks that have a healthy amount of comments

Reviewer: maybe explain what notebooks are or remove it?

Author: I thought you were planning on adding a chapter before this about RMarkdowns and notebooks?

Author: We should definitely link that chapter when it exists.

numbers<-  c(1, 4,5,10)

# Good!
numbers <- c(1, 4, 5, 10)
```

Here's some style guides you can use:

- [Google's R Style Guide](https://google.github.io/styleguide/Rguide.html)
- [Tidyverse Style Guide](https://style.tidyverse.org/index.html)

[R packages like styler](https://www.tidyverse.org/blog/2017/12/styler-1.0.0/) can automatically style code for you.

## Efficient

Reproducible code is efficient code. Efficiency helps reproducibility in that code that takes up less resources and is not redundant can be re-run and debugged more easily.

### Doesn’t use up more computational resources than necessary

R is not meant to be a fast language. R code can be computationally costly if it's written in the wrong way.

You can identify what parts of your R code are the slowest or otherwise computationally costly by using the [`profvis`](https://rstudio.github.io/profvis/) and [`bench`](https://bench.r-lib.org/) packages.

One popular example is R loops which can be particularly slow in R. Note that this doesn't mean you shouldn't ever use loops or other items in R, just that you should be aware that some items in R are particularly slower than others.

**R Loop alternatives:**

> - [Using lapply](https://statisticsglobe.com/avoid-for-loop-in-r)

Reviewer: ```suggestion
- [Using `apply` functions (an older option that comes with the basic installation of R)](https://statisticsglobe.com/avoid-for-loop-in-r)
```

Reviewer: being more general might be useful?

> - [Using the purrr package](https://purrr.tidyverse.org/)

Reviewer: ```suggestion
- [Using the `purrr` package](https://purrr.tidyverse.org/)
```

- [Advanced R discusses alternative strategies](https://adv-r.hadley.nz/perf-improve.html?q=perfor#avoid-copies)
- [The across function](https://dplyr.tidyverse.org/reference/across.html)

> 

Reviewer: We also have a summary of each of these here on apply, across and purrr (mostly using sapply since the students in that class tend to want to work with dataframes)
https://jhudatascience.org/intro_to_r/modules/Functions/Functions.html#18

ottrpal::include_slide("https://docs.google.com/presentation/d/1MNHf8JpolaEP_vQ_kB-1xRBF9wo3haCArRu117hBoHA/edit#slide=id.g221fb1e0594_0_0")
```

## Reproducibility means we don't need to reinvent the wheel!

> When you realize something that you need done, you should first use Google and look on GitHub and StackOverflow for is someone else has written something that works really well. Where at all possible, borrow good code and attribute the author -- no need to reinvent the wheel. As we said previously, R has a great community of users who are constantly creating new and great code, often in the form of packages that are ready for you to install and use.

Reviewer: ```suggestion
When you realize something that you need done, you should first use Google and look on GitHub and StackOverflow to see if someone else has written something that works really well. Where at all possible, borrow good code and attribute the author -- no need to reinvent the wheel. As we said previously, R has a great community of users who are constantly creating new and great code, often in the form of packages that are ready for you to install and use.
```

_This is the beauty of reproducibility, if you or someone else makes great code, it can not only be re-run but it can be repurposed!_

### Tips for choosing packages to use:

> - Does the package have easy to use documentation to help guide you on how to use it properly? If the package has underdeveloped documentation it may be difficult for you or other to understand and use properly.

Reviewer: - Does the package have easy to use documentation to help guide you on how to use it properly? If the package has underdeveloped documentation it may be difficult for you or others to understand and use properly.

- Is the code actively being developed or maintained? Packages that are no longer being maintained will likely deprecate more quickly rendering them unusable in the future.
- Is it a package that is commonly recognized by the community? Well recognized packages will be easier for others to comment on and help you with.

## The importance of iterative work

> We've mentioned previously that reproducibility is iterative work. This way of working refers to code work as much as anything else. You won't ever right perfect code on the first try, instead aim for each chunk to work one step at a time. Once it is working, take a break (perhaps until the next day), then return to it and look for ways to polish it and make it more efficient.

Reviewer: ```suggestion
We've mentioned previously that reproducibility is iterative work. This way of working refers to code work as much as anything else. You won't ever write perfect code on the first try, instead aim for each chunk to work one step at a time. Once it is working, take a break (perhaps until the next day), then return to it and look for ways to polish it and make it more efficient.
```

Author: Lolz

Code style helps make code readable. Appropriate spacing, punctuation, and grammar are not always essential for getting a message across, but it can certainly b3 dis-tRaCTIng to readers if conventions aren't followed.

**Basic Example:**
```{r}
# Bad: Should use <- and have a variable name that is informative
x = c(1, 4, 5, 10)

# Bad: Irregular spacing is distracting

> numbers<-  c(1, 4,5,10)

Reviewer: love it! :)

- [Using lapply](https://statisticsglobe.com/avoid-for-loop-in-r)

> - [Using the purrr package](https://purrr.tidyverse.org/)

Reviewer: maybe we should move this one down to be the last listed? I like the purrr package but beginners tend to find it challenging.

Author: Fair! Done.
